src/snowflake/snowpark/dataframe.py

```diff
@@ -2245,14 +2256,30 @@ def natural_join(
             --------------------
             <BLANKLINE>
         """
-        join_type = kwargs.get("join_type") or how
+        join_type = create_join_type(kwargs.get("join_type") or how or "inner")
         join_plan = Join(
             self._plan,
             right._plan,
-            NaturalJoin(create_join_type(join_type or "inner")),
+            NaturalJoin(join_type),
             None,
             None,
         )
+
+        stmt = self._session._ast_batch.assign()
+        ast = with_src_position(stmt.expr.sp_dataframe_natural_join, stmt)
+        self.set_ast_ref(ast.lhs)
+        right.set_ast_ref(ast.rhs)
+        if isinstance(join_type, Inner):
+            ast.join_type.sp_join_type__inner = True
+        elif isinstance(join_type, LeftOuter):
+            ast.join_type.sp_join_type__left_outer = True
+        elif isinstance(join_type, RightOuter):
+            ast.join_type.sp_join_type__right_outer = True
+        elif isinstance(join_type, FullOuter):
+            ast.join_type.sp_join_type__full_outer = True
+        else:
+            raise ValueError(f"Unsupported join type {join_type}")
+
         if self._select_statement:
             select_plan = self._session._analyzer.create_select_statement(
```

Comment:
I think the problem is here. Can you check that the return value has your AST?

Reply:
My suspicion is that you need to call `create_selet_statement({current args}, stmt)`

Reply:
Hmm. This location is in `Dataframe.natural_join`, which currently passes the tests. I'm having issues with `Dataframe.join`, where I actually made a change similar in spirit to this suggestion.

Reply:
As we discovered in person, the problem was with the `Dataframe.show()` calls.